test: refactor: various (de)serialization helpers cleanups/improvements

theStack opened the pull request:
There are still many functional tests that perform conversions from a hex-string to a message object (deserialization) manually. This PR identifies all those instances and replaces them with a newly introduced helper `tx_from_hex`.

Instances were found via
* `git grep "deserialize.*BytesIO"`

and some of them manually, when it were not one-liners.

Further, the helper `ToHex` was removed and simply replaced by `.serialize().hex()`, since now both variants are in use (sometimes even within the same test) and using the helper doesn't really have an advantage in readability. (see discussion https://github.com/bitcoin/bitcoin/pull/22257#discussion_r652404782)

theStack committed:
test: introduce `tx_from_hex` helper for tx deserialization

`FromHex` is mostly used for transactions, so we introduce a
shortcut `tx_from_hex` for `FromHex(CTransaction, hex_str)`.

theStack committed:
test: remove `ToHex` helper, use .serialize().hex() instead

klementtan reviewed (commented):
Code review ACK 82449a55f16b7a2897ccef999c3f807ec1a1e591

Verified that `.serialize().hex()` to `ToHex(obj)` and `obj.deserialize(BytesIO(hex_str_to_bytes(hex_string)))` to `FromHex(obj, hex_string)` are the only changes made.

Also manually checked with `rg '.deserialize' -g '*.py'` and `rg '\.hex\(\)' -g '*.py` that there are no other places to refactor.

fanquake commented:
Concept ACK

practicalswift commented:
Concept ACK: nice cleanup!

MarcoFalke commented on test/functional/feature_cltv.py:
```diff
@@ -166,7 +167,7 @@ def run_test(self):
                     'reject-reason': expected_cltv_reject_reason,
                 }],
-                self.nodes[0].testmempoolaccept(rawtxs=[spendtx.serialize().hex()], maxfeerate=0),
+                self.nodes[0].testmempoolaccept(rawtxs=[ToHex(spendtx)], maxfeerate=0),
```
Is there a reason to do this replacement? Now that python supports bin->hex by just appending `.hex()`, the helper isn't really needed anymore.

Personally I think both versions are fine and there isn't really a way to "protect" new code from using one or the other.

theStack replied:
Agree that the gain can be questioned -- I was also hesitating about this replacement, but then thought symmetry (i.e. why get rid of most deserialize() calls, but let the serialize() calls in) and the fact that it's shorter may be worth it.

Some tests already imported and used ToHex, but at some places still called serialize().hex(). At least mixing within the same test should be avoided IMHO, as it only creates confusion.

Maybe in the long-term, removing `ToHex` completely would be an alternative?

MarcoFalke replied:
Then my preference would be to remove `ToHex` to avoid future pull requests changing it back (or forth)

maflcko reviewed (commented):
Not sure about the long term goal of the `ToHex` replacement.

theStack commented:
Force-pushed with a change in the second commit: `ToHex` is removed and all instances are replaced with `.serialize().hex()`, as suggested [in this discussion by MarcoFalke](https://github.com/bitcoin/bitcoin/pull/22257#discussion_r652469155). Also adapted PR title and description accordingly.

practicalswift commented on test/functional/test_framework/messages.py:
```diff
@@ -195,10 +195,6 @@ def FromHex(obj, hex_string):
     obj.deserialize(BytesIO(hex_str_to_bytes(hex_string)))
     return obj
 
-# Convert a binary-serializable object to hex (eg for submission via RPC)
```
Non-blocking suggestion: Could add a comment to `FromHex` describing that the recommended inverse is `obj.serialize().hex()` (instead of a hypothetical `ToHex`) :)

theStack replied:
Good idea, I added a commit with your suggestion!

brunoerg reviewed (approved):
tACK 6665fec39e8e42e05a7b602a24f0f6d312cf3991

Nice improv! 

Ran test runner and every changed test several times on MacOS 11.3.

theStack committed:
test: doc: improve doc for `from_hex` helper (mention `to_hex` alternative)

theStack commented:
Added a commit https://github.com/bitcoin/bitcoin/pull/22257/commits/02d2f961c1c3977c898a1180b6187cadd2f1080f to improve the documentation of `FromHex` by mentioning the recommended reverse operation ([as suggested by practicalswift](https://github.com/bitcoin/bitcoin/pull/22257#discussion_r653036925)) and changing to docstring format.

practicalswift commented:
cr ACK 18b780b6ce2b42245f067e922af5552bd8931be3

theStack commented:
Rebased on master.

theStack commented on test/functional/mempool_accept.py:
```diff
@@ -91,8 +90,7 @@ def run_test(self):
         ))['hex']
-        tx = CTransaction()
-        tx.deserialize(BytesIO(hex_str_to_bytes(raw_tx_0)))
+        tx = FromHex(CTransaction(), raw_tx_0)
```
Good idea, done.

theStack commented:
Added another commit to introduce a `tx_from_hex` helper, as suggested by MarcoFalke (https://github.com/bitcoin/bitcoin/pull/22257#discussion_r654817481).
Tempted to squash this with the first commit to minimize reviewing burden, OTOH both replacing code with an existing helper _and_ introducing introducing a new one seems a bit much for a single commit. Opinions?

MarcoFalke commented:
I have a slight preference to squash to avoid repeatedly touching the same lines

theStack committed:
scripted-diff: test: rename `FromHex` to `from_hex`

-BEGIN VERIFY SCRIPT-
sed -i 's/\<FromHex\>/from_hex/g' $(git grep -l FromHex)
-END VERIFY SCRIPT-

Co-authored-by: MarcoFalke <falke.marco@gmail.com>

theStack commented:
Restructured everything a bit (sorry for confusing to all who already reviewed!):
* the first commit immediately introduces a helper `tx_from_hex`, which is used for instances where deserialization was done manually, and also for instances which used `FromHex` already
* second commit gets rid of `ToHex`
* third commit renames `FromHex` to `from_hex` via scripted diff (to follow coding guidelines / PEP8 naming)
* fourth commit adds `from_hex` documentation

MarcoFalke commented:
I've rewritten the scripted diff and it seems there is a silent merge conflict?

```
$ grd 8ed7dbd386562368e864fce23722676cd3637854 HEAD
1:  8c127a86be = 1:  8c127a86be test: introduce `tx_from_hex` helper for tx deserialization
2:  4223f18a8c = 2:  4223f18a8c test: remove `ToHex` helper, use .serialize().hex() instead
3:  8ed7dbd386 ! 3:  662223f4ba scripted-diff: test: rename `FromHex` to `from_hex`
    @@ Commit message
         scripted-diff: test: rename `FromHex` to `from_hex`
     
         -BEGIN VERIFY SCRIPT-
    -    git ls-files -- test/functional | xargs sed -i 's/FromHex/from_hex/g'
    +    sed -i 's/\<FromHex\>/from_hex/g' $(git grep -l FromHex)
         -END VERIFY SCRIPT-
     
      ## test/functional/feature_utxo_set_hash.py ##

MarcoFalke commented:
review ACK 446fd1792ae3e7d0917df771bc0ccfa7c3c3b8f6 🏵

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 446fd1792ae3e7d0917df771bc0ccfa7c3c3b8f6 🏵
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiqPAv8DOhtbQZoGRJyrizrhOHso5pM2dbArc/E6N3UEPT/FpHLBHuZiEvgbKeA
CmPU4f9QQKrW4pLMIdF7i/R94Lwo7pNulA0697LyZMSypRSMZiLIq6WBSaY7U6+o
9Ao7AOf9gzBBhqs9qWeFrk0dSK8uaEcPMhe+poDOvzr+8BGyf49amhySCUFo9CRQ
gsjhPVt2/cEfX7Chg+iUVotW35t5lLid8NUBpelumapeeflZ0943cguv6QSy6IZ8
shtdjptvn/VFMi8nhqeBPTspKrQfY29cF3h0DBjPxFqWdERgdBhdNetG3wIGHTzQ
Hg/yf1FfsBs7Sk9MFDE+HW+u1i8LhmzCFkP3rRXI8qhY7RGCWKe4NFSJUJLi+4JB
CpPtsPSjJkomj6rqec34GqWDuGzkxOqpWM5j++xpCp46pzA4zkdtICecnp3VCcYv
8vGpeU2K6x2tKX7ZiTMSKZ3m7XqoF2x9RcP/NP2kWDVzdA4f184VXoJVT0DHRF5Z
bpeDtV39
=U71S
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `a3b9802ea42e8e0ba75b10e75c2b16701bc96227df1740b76ceb4b2ee5a11dda  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108a3b9802ea42e8e0ba75b10e75c2b16701bc96227df1740b76ceb4b2ee5a11ddaf010be79f22bf6dd8d157a6bf685631ee81e08fff0101262dfe246556c1c5ab0c032f463d4b308f10460d05b0bf0089099b32a880345c10083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01038855c9c37a05e590da9dd5f4a78f19708f10460d05b0bf0083a54fd8208ec69480083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff01047760e24fa47138bc16e7dbb7e48737908f10460d05b0bf00802c8316d2652a6800083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010fa903a6254a55020537d52c92180b93108f10460d05b0bf00843047f7ad51083780083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

theStack commented:
> I've rewritten the scripted diff and it seems there is a silent merge conflict?
> 
> ```
> $ grd 8ed7dbd386562368e864fce23722676cd3637854 HEAD
> 1:  8c127a86be = 1:  8c127a86be test: introduce `tx_from_hex` helper for tx deserialization
> 2:  4223f18a8c = 2:  4223f18a8c test: remove `ToHex` helper, use .serialize().hex() instead
> 3:  8ed7dbd386 ! 3:  662223f4ba scripted-diff: test: rename `FromHex` to `from_hex`
>     @@ Commit message
>          scripted-diff: test: rename `FromHex` to `from_hex`
>      
>          -BEGIN VERIFY SCRIPT-
>     -    git ls-files -- test/functional | xargs sed -i 's/FromHex/from_hex/g'
>     +    sed -i 's/\<FromHex\>/from_hex/g' $(git grep -l FromHex)
>          -END VERIFY SCRIPT-
>      
>       ## test/functional/feature_utxo_set_hash.py ##
> ```

Oops, I completely forgot that we also have Python scripts outside of test/functional that use helpers from the test framework and are thus affected by the changes (namely `contrib/signet/miner`).

Force-pushed with changes including it (affects all but the last commit), used your scripted-diff and add you as co-author.

(Not directly connected to this PR, but it would probably make sense to call the signet mining tool in one of the functional tests, to let it run through CI?)

MarcoFalke commented:
review re-ACK bdb8b9a347e68f80a2e8d44ce5590a2e8214b6bb 😁

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review re-ACK bdb8b9a347e68f80a2e8d44ce5590a2e8214b6bb 😁
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjrswv/bCHcA6Zpk9Gv9fkRzBtmSfbuLE2ifDg+RGN2prKClfrlim2Kx9LhjTK+
BoeGGWwMAUdg+SHdNi5RMhmiRrSB+4JMtxw3TY6COBjP/QYdmzyySle/xNGwAoov
0CGZsEZaqiC5QPFJe6pLZJMRJET8iJR9gLaNNkWbEkUllqkjdvaIuGCJjKC4k1ld
epdUxh+yxg5U/FPMgSR5jPew1GBMpb6XPk0YSJ2yPimCBlSeaIH9X1XlYNix1WPv
tn6rCN4ZUkufPqb8x0R//y/x3T0lMOgjOsnODR5BZtUdV0P/GK7ZTRa9zBtz/Xa4
D7somqqh5TKOBvNJounyLS1FO3zXBoE6Kbr5xZQinob7fpTlEf9nvS9yWNtE+tyZ
em+74lyfdHMiBBWmlEIIiHaOrrG6pPOXQ6zn/GSEkrlHjQLtLJ2a+YoAzO6hc6Af
RUA/NKPHz8NPHcOzpbtPTs+sYloLrtbp8O5dUsgyZ7JWTT0hAEwIpahQsEoi9j4b
lnHvly/x
=rKMS
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `1cd7510533208f2380d21b0afee7343c273d929ac88f1e044e3c4c7724daab3b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401081cd7510533208f2380d21b0afee7343c273d929ac88f1e044e3c4c7724daab3bf0105ddc567b4c415281c3a797bf9619947508fff01044467892775bcd4f4094b224a6b230c808f10460d2dd09f0085e721027dad80c040083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0108f00763e470b1435d3633c45624977fc08f10460d2dd09f008d82e99d061050e670083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010b470e54b1f18d54840551bed01b571f808f10460d2dd09f008128a25a5e61de16d0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010f058032c7a8bd94a23b913693b64325e08f1203b41e7b3a00719bff6d82fbb80de5f035190986060d817d4886881524dccbd7208f10460d2dd09f008337f657053eea8150083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented:
> (Not directly connected to this PR, but it would probably make sense to call the signet mining tool in one of the functional tests, to let it run through CI?)

Agree
